File: sample-videochat-webrtc/sample-videochat-webrtc/Classes/ConnectionModule/ConnectionModule.m
//
//  Connection.m
//  sample-videochat-webrtc
//
//  Created by Injoit on 06.10.2020.
//  Copyright © 2020 QuickBlox Team. All rights reserved.
//

#import "ConnectionModule.h"
#import "Profile.h"
#import <arpa/inet.h>

@interface ConnectionModule ()

@property (nonatomic, assign) BOOL isProcessing;
@property (nonatomic, assign) SCNetworkReachabilityRef reachability;
@property (nonatomic, assign, readonly) BOOL isNetworkLost;

@property (nonatomic, assign) BOOL activeCall;

@end

@interface ConnectionModule (ChatConnection) <QBChatDelegate>
@end

@implementation ConnectionModule

- (SCNetworkReachabilityRef)reachability {
    if (_reachability) {
        return _reachability;
    }
    struct sockaddr_in address;
    bzero(&address, sizeof(address));
    address.sin_len = sizeof(address);
    address.sin_family = AF_INET;
    
    _reachability =
    SCNetworkReachabilityCreateWithAddress(kCFAllocatorDefault, (const struct sockaddr *)&address);
    return _reachability;
}

- (BOOL)isNetworkLost {
    SCNetworkReachabilityFlags flags;
    if (SCNetworkReachabilityGetFlags(self.reachability, &flags)) {
        
        return (flags & kSCNetworkReachabilityFlagsReachable) == 0;
    }
    return NO;
}

- (BOOL)tokenHasExpired {
    return QBSession.currentSession.tokenHasExpired;
}

- (BOOL)established {
    BOOL connected = NO;
    if (self.tokenHasExpired == NO && QBChat.instance.isConnected) {
        connected = YES;
    }

    if (connected == NO) {
        [self establishConnection];
    }
    return connected;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        self.isProcessing = NO;
        self.activeCall = NO;
        QBSettings.autoReconnectEnabled = YES;
        QBSettings.networkIndicatorManagerEnabled = YES;
        [QBChat.instance addDelegate:self];
    }
    return self;
}

- (void)dealloc {
    [QBChat.instance removeDelegate:self];
}

//MARK: - Actions
- (void)activateAutomaticMode {
    [self establishConnection];
}

- (void)activateCallMode {
    self.activeCall = YES;
}

- (void)deactivateCallMode {
    self.activeCall = NO;
    if (UIApplication.sharedApplication.applicationState == UIApplicationStateActive) {
        return;
    }
}

- (void)establishConnection {
    Profile *profile = [[Profile alloc] init];
    [self connectWithId:profile.ID
               password:profile.password];
}

- (void)breakConnectionWithCompletion:(nonnull void (^)(void))completion {
    if (self.isProcessing) {
        return;
    }
    self.isProcessing = YES;
    
    if (QBChat.instance.isConnected == NO) {
        [self logoutWithCompletion:completion];
        return;
    }
    __weak __typeof(self)weakSelf = self;
    [QBChat.instance disconnectWithCompletionBlock:^(NSError * _Nullable error) {
        [weakSelf logoutWithCompletion:completion];
    }];
}

//MARK: - Internal
- (void)logoutWithCompletion:(nonnull void (^)(void))completion {
    __weak __typeof(self)weakSelf = self;
    [QBRequest logOutWithSuccessBlock:^(QBResponse * _Nonnull response) {
        weakSelf.isProcessing = NO;
        completion();
    } errorBlock:^(QBResponse * _Nonnull response) {
        weakSelf.isProcessing = NO;
        completion();
    }];
}

- (void)connectWithId:(NSUInteger)id password:(NSString *)password {
    if (QBChat.instance.isConnected || QBChat.instance.isConnecting) {
        return;
    }
    [QBChat.instance connectWithUserID:id
                              password:password
                            completion:nil];
    return;
}

- (void)disconnect {
    if (QBChat.instance.isConnected == NO) {
        return;
    }
    [QBChat.instance disconnectWithCompletionBlock:nil];
}

@end

@implementation ConnectionModule (ChatConnection)

- (void)chatDidConnect {
    if (self.onConnect) { self.onConnect(); }
}

- (void)chatDidNotConnectWithError:(NSError *)error {
    if (error == nil) {
        return;
    }
    if (self.onDisconnect) {
        self.onDisconnect(self.isNetworkLost);
    }
}

- (void)chatDidDisconnectWithError:(NSError *)error {
    if (error == nil) {
        return;
    }
    if (self.onDisconnect) {
        self.onDisconnect(self.isNetworkLost);
    }
}

- (void)chatDidReconnect {
    if (self.onConnect) { self.onConnect(); }
}

@end

File: sample-videochat-webrtc-swift/sample-videochat-webrtc-swift/Components/Reachability/ConnectionModule.swift
//
//  ConnectionModule.swift
//  sample-videochat-webrtc-swift
//
//  Created by Injoit on 02.04.2021.
//  Copyright © 2021 QuickBlox. All rights reserved.
//

import UIKit
import Quickblox
import QuickbloxWebRTC
import SystemConfiguration

typealias CompletionAction = (() -> Void)
typealias DisconnectionAction = ((Bool?) -> Void)

/// The module is responsible for automatic connection establishment with QuickBlox.
///
/// The connection depends on the application state and call activity.
/// The establishing connection process can start from the authorization if the "QBSession" token was expired.
class ConnectionModule: NSObject {
    //MARK: - Properties
    /// Determining the connection state.
    ///
    /// Calling this property starts the connection process when it's not established.
    var established: Bool {
        var connected = false
        if tokenHasExpired == false, QBChat.instance.isConnected {
            connected = true
        }
        
        if connected == false {
            establishConnection()
        }
        return connected
    }
    /// The authorization process running.
    private(set) var isProcessing: Bool = false
    /// Determining the "QBSession" token state.
    var tokenHasExpired: Bool {
        return QBSession.current.tokenHasExpired
    }

    /// Called when authorization complete.
    var onAuthorize: CompletionAction?
    /// Called when the connection was established or re-established.
    var onConnect: CompletionAction?
    /// Called when connection lost.
    var onDisconnect: DisconnectionAction?
    
    private var isReachability: SCNetworkReachability? {
        return reachability()
    }
    private var isNetwork: Bool? {
        var flags: SCNetworkReachabilityFlags = []
        guard let reachability = isReachability,
              SCNetworkReachabilityGetFlags(reachability, &flags) else {
            return nil
        }
        return flags.contains(.reachable)
    }
    
    private var activeCall: Bool = false
    
    
    //MARK: - Life Cycle
    override init() {
        super.init()
        
        QBSettings.autoReconnectEnabled = true
        QBSettings.networkIndicatorManagerEnabled = true
        QBChat.instance.addDelegate(self)
    }
    
    deinit {
        QBChat.instance.removeDelegate(self)
    }
    
    // MARK: - Public Methods
    func reachability() -> SCNetworkReachability? {
        var zeroAddress = sockaddr_in()
        zeroAddress.sin_len = UInt8(MemoryLayout<sockaddr_in>.size)
        zeroAddress.sin_family = sa_family_t(AF_INET)
        
       let isReachability = withUnsafePointer(to: &zeroAddress, {
            $0.withMemoryRebound(to: sockaddr.self, capacity: 1) {
                SCNetworkReachabilityCreateWithAddress(nil, $0)
            }
        })
        return isReachability
    }

    /// Activating the automatic connection and disconnection process when the application state change.
    ///
    /// Calling this method starts the connection process when it's not established.
    func activateAutomaticMode() {
        establishConnection()
    }
    
    /// Prevents breaking connection when the application state change.
    func activateCallMode() {
        activeCall = true
    }
    
    /// Break connection when application state is inactive.
    func deactivateCallMode() {
        activeCall = false
    }
    
    /// Establishes a connection with the Quickblox.
    func establishConnection() {
        let profile = Profile()
        guard profile.isFull else {
            return
        }
        self.connect(withId: profile.ID, password: profile.password)
    }
    
    /// Disconnects and unauthorize from the Quickblox.
    func breakConnection(withCompletion completion:@escaping CompletionAction) {
        if isProcessing {
            return
        }
        isProcessing = true
        
        if QBChat.instance.isConnected == false {
            logout(withCompletion: completion)
            return
        }
        
        QBChat.instance.disconnect { [weak self] (error) in
            self?.logout(withCompletion: completion)
        }
    }
    
    //MARK: - Internal
    private func logout(withCompletion completion:@escaping CompletionAction) {
        QBRequest.logOut(successBlock: { [weak self] response in
            self?.isProcessing = false
            completion()
        }) {  [weak self] response in
            self?.isProcessing = false
            completion()
        }
    }

    private func connect(withId userID:UInt, password: String) {
        if QBChat.instance.isConnected || QBChat.instance.isConnecting {
            return
        }
        QBChat.instance.connect(withUserID: userID, password: password, completion: nil)
        return
    }
    
    private func disconnect() {
        if QBChat.instance.isConnected == false {
            return
        }
        QBChat.instance.disconnect(completionBlock: nil)
    }
}

//MARK: - QBChatDelegate
extension ConnectionModule: QBChatDelegate {
    func chatDidConnect() {
        onConnect?()
    }
    
    func chatDidNotConnectWithError(_ error: Error) {
        onDisconnect?(isNetwork)
    }
    
    func chatDidDisconnectWithError(_ error: Error?) {
        guard error != nil else {
            return
        }
        onDisconnect?(isNetwork)
    }
    
    func chatDidReconnect() {
        onConnect?()
    }
}